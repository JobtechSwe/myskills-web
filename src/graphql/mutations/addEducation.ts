import { InMemoryCache } from 'apollo-cache-inmemory'
import { Education } from '../../types'
import gql from 'graphql-tag'

export const GET_EDUCATIONS_CLIENT = gql`
  query getEducations {
    educations @client
  }
`

export const addEducation = (
  _: any,
  { education }: { education: Education },
  { cache }: { cache: InMemoryCache }
): Education => {
  const { educations }: any = cache.readQuery({
    query: GET_EDUCATIONS_CLIENT,
  })

  const withoutDuplicates = (educations: Education[]): Education[] =>
    educations.filter((e: Education) => e.taxonomyId !== education.taxonomyId)

  const updatedEducations = [...withoutDuplicates(educations), education]

  cache.writeQuery({
    query: GET_EDUCATIONS_CLIENT,
    data: { educations: updatedEducations },
  })

  return education
}
