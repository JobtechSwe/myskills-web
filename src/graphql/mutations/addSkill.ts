import { InMemoryCache } from 'apollo-cache-inmemory'
import { Skill } from '../../types'
import gql from 'graphql-tag'

export const GET_SKILLS_CLIENT = gql`
  query getSkills {
    skills @client
  }
`

export const addSkill = (
  _: any,
  { skill }: { skill: Skill },
  { cache }: { cache: InMemoryCache }
): Skill => {
  const { skills }: any = cache.readQuery({
    query: GET_SKILLS_CLIENT,
  })

  const withoutDuplicates = (skills: Skill[]): Skill[] =>
    skills.filter((s: Skill) => s.taxonomyId !== skill.taxonomyId)

  const updatedSkills = [...withoutDuplicates(skills), skill]

  cache.writeQuery({
    query: GET_SKILLS_CLIENT,
    data: { skills: updatedSkills },
  })

  return skill
}
