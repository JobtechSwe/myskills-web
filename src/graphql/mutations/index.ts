export * from './addExperience'
export * from './addLanguage'
export * from './addSkill'
export * from './addEducation'
export * from './removeEducation'
export * from './removeExperience'
