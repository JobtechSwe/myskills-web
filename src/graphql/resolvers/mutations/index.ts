export * from './addLanguage'
export * from './addSkill'
export * from './addEducation'
export * from './addExperience'
export * from './addTrait'
export * from './addWhoAmI'
export * from './createOccupation'
export * from './updateContactInformation'
export * from './removeEducation'
export * from './removeExperience'
export * from './removeOccupation'
export * from './removeSkill'
export * from './removeTrait'
export * from './updateEducation'
export * from './updateExperience'
export * from './uploadImage'
