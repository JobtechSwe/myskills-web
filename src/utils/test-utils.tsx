import * as React from 'react'
import ApolloClient from 'apollo-client'
import { ApolloProvider } from 'react-apollo-hooks'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { MockLink, MockedProvider } from 'react-apollo/test-utils'
import { ThemeProvider } from 'emotion-theming'
import { render as rtlRender } from 'react-testing-library'
import { theme } from 'theme'

export const render = (children: React.ReactNode, mocks?: any[]) => {
  const client = new ApolloClient({
    cache: new InMemoryCache({ addTypename: true }),
    link: new MockLink(mocks || []),
  })

  return rtlRender(
    <MockedProvider addTypename={true} mocks={mocks}>
      <ApolloProvider client={client}>
        <ThemeProvider theme={theme}>{children}</ThemeProvider>
      </ApolloProvider>
    </MockedProvider>
  )
}
