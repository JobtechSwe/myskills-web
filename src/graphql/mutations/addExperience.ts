import { InMemoryCache } from 'apollo-cache-inmemory'
import { Experience } from '../../types'
import { storageHelper } from '../../utils/helpers'
import gql from 'graphql-tag'

export const GET_EXPERIENCES_CLIENT = gql`
  query getExperiences {
    experiences @client {
      name
      taxonomyId
      years
    }
  }
`

export const addExperience = (
  _: any,
  { experience }: { experience: Experience },
  { cache }: { cache: InMemoryCache }
): Experience => {
  const { experiences }: any = cache.readQuery({
    query: GET_EXPERIENCES_CLIENT,
  })

  const withoutDuplicates = (exp: Experience[]): Experience[] =>
    exp.filter((e: Experience) => e.taxonomyId !== experience.taxonomyId)

  const updatedExperiences = [...withoutDuplicates(experiences), experience]

  cache.writeQuery({
    query: GET_EXPERIENCES_CLIENT,
    data: { experiences: updatedExperiences },
  })

  storageHelper.set({
    type: 'experiences',
    data: updatedExperiences,
  })

  return experience
}
