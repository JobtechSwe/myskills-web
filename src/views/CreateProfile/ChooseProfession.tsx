import React, { useState } from 'react'
import { useQuery, useMutation } from 'react-apollo-hooks'
import gql from 'graphql-tag'
import {
  OntologyType,
  OntologyConceptResponse,
} from '../../generated/myskills.d'
import { RouteComponentProps } from '@reach/router'
import Grid from '../../components/Grid'
import Input from '../../components/Input'
import IllustrationHeader from '../../components/IllustrationHeader'
import suitcaseIllustration from '../../assets/illustrations/suitcase.svg'
import ListItem from '../../components/ListItem'
import { SearchList } from '../../components/List'
import styled from '@emotion/styled'
import { css, Global } from '@emotion/core'
import ChosenOccupation from '../../components/ChosenOccupation'
import Downshift from 'downshift'
import { highlightMarked } from '../../utils/helpers'
import RegistrationLayout from '../../components/Layout/RegistrationLayout'

const SearchInput = styled(Input)`
  width: 100%;
`

export const GET_ONTOLOGY_CONCEPTS = gql`
  query ontologyConcepts($filter: String!, $type: OntologyType) {
    ontologyConcepts(params: { filter: $filter, type: $type }) {
      id
      term
      type
    }
  }
`

export const CREATE_OCCUPATION_CLIENT = gql`
  mutation createOccupationClient($occupation: OccupationInput!) {
    createOccupationClient(occupation: $occupation) @client {
      term
      experience {
        years
      }
    }
  }
`

export const CREATE_OCCUPATION_API = gql`
  mutation createOccupationApi($occupation: OccupationInput!) {
    createOccupation(occupation: $occupation) {
      term
      experience {
        years
      }
    }
  }
`

/*
  (always: true) is currently unreliable but Apollo reports bug is fixed in 2.6.0
  https://github.com/apollographql/apollo-client/issues/4636#issuecomment-480307041
*/
export const IS_LOGGED_IN = gql`
  query isLoggedIn {
    isLoggedIn @client(always: true)
  }
`

const ChooseProfession: React.FC<RouteComponentProps> = () => {
  const [query, setQuery] = useState('')
  // TODO: Use debounce, skipping for now because of complications in tests
  const { data: isLoggedIn } = useQuery(IS_LOGGED_IN, {
    fetchPolicy: 'network-only',
  })

  const createOccupation = useMutation(
    isLoggedIn.isLoggedIn ? CREATE_OCCUPATION_API : CREATE_OCCUPATION_CLIENT
  )

  const { data, error } = useQuery(GET_ONTOLOGY_CONCEPTS, {
    variables: {
      filter: query,
      type: OntologyType.Occupation,
    },
    skip: !query,
  })

  if (error) {
    return <div>Error...</div>
  }

  return (
    <RegistrationLayout headerText="YRKE" nextPath="kompetenser" step={1}>
      <Grid alignContent="start">
        <IllustrationHeader
          imageAltTag="Resväska"
          imageFirst={true}
          imageSource={suitcaseIllustration}
          title="Vad vill du jobba med?"
        />
        <Global
          styles={css`
            strong {
              font-weight: 700;
              text-transform: capitalize;
            }
          `}
        />
        <Downshift
          itemToString={item => (item ? item.term : '')}
          onChange={occupation => {
            createOccupation({
              variables: {
                occupation: {
                  term: occupation.term,
                  experience: null,
                },
              },
            })
            setQuery('')
          }}
        >
          {({
            getInputProps,
            getItemProps,
            getMenuProps,
            isOpen,
            inputValue,
            highlightedIndex,
          }) => (
            <div>
              <SearchInput
                {...getInputProps({
                  name: 'search',
                  placeholder: 'Yrkesroll eller yrkesområde',
                  onChange: (event: React.ChangeEvent<HTMLInputElement>) =>
                    setQuery(event.target.value),
                  value: query,
                })}
              />

              {data && data.ontologyConcepts && (
                <SearchList isOpen={isOpen} {...getMenuProps()}>
                  {isOpen
                    ? data.ontologyConcepts
                        .filter(
                          (item: OntologyConceptResponse) =>
                            !inputValue ||
                            item.term
                              .toLowerCase()
                              .includes(inputValue.toLowerCase())
                        )
                        .map((item: OntologyConceptResponse, index: number) => (
                          <ListItem
                            bg={
                              highlightedIndex === index
                                ? 'seashellPeach'
                                : 'white'
                            }
                            key={item.id}
                            px="medium"
                            {...getItemProps({
                              key: item.id,
                              index,
                              item,
                              dangerouslySetInnerHTML: highlightMarked(
                                inputValue,
                                item.term
                              ),
                            })}
                          />
                        ))
                    : null}
                </SearchList>
              )}
            </div>
          )}
        </Downshift>
        <ChosenOccupation />
      </Grid>
    </RegistrationLayout>
  )
}

export default ChooseProfession
