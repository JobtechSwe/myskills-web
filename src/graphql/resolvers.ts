import { GET_EXPERIENCES_CLIENT } from '../views/CreateProfile/AddExperience'

const resolvers = {
  Mutation: {
    addExperience: (
      _: any,
      { experience }: any,
      { cache }: { cache: any }
    ): any => {
      const { experiences } = cache.readQuery({
        query: GET_EXPERIENCES_CLIENT,
      })
      const expericensWithoutDuplicates = experiences.filter(
        (e: any) => e.taxonomyId !== experience.taxonomyId
      )

      const updatedExperiences = [...expericensWithoutDuplicates, experience]

      cache.writeQuery({
        query: GET_EXPERIENCES_CLIENT,
        data: { experiences: updatedExperiences },
      })

      return experience
    },
  },
}

export default resolvers
