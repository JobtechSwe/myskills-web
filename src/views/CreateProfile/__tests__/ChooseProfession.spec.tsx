import * as React from 'react'
import ChooseProfession, { GET_ONTOLOGY_CONCEPTS } from '../ChooseProfession'
import { render } from '../../../utils/test-utils'
import { wait, fireEvent } from 'react-testing-library'
import { OntologyType } from '../../../generated/myskills.d'
import { GET_OCCUPATIONS_CLIENT } from '../../../graphql/resolvers/mutations'

describe('views/ChooseProfession', () => {
  let withResultsMock: any

  beforeEach(() => {
    withResultsMock = [
      {
        request: {
          query: GET_ONTOLOGY_CONCEPTS,
          variables: {
            filter: 'Systemutvecklare',
            type: OntologyType.Occupation,
          },
        },
        result: {
          data: {
            __typename: 'OntologyConceptsResult',
            ontologyConcepts: [
              {
                name: 'Systemutvecklare',
                id: 'abc',
                type: 'occupation-name',
                __typename: 'OntologyConceptResult',
              },
            ],
          },
        },
      },
    ]
  })

  it('renders empty result', async () => {
    const withResultsMock = [
      {
        request: {
          query: GET_OCCUPATIONS_CLIENT,
        },
        result: {
          data: {
            occupations: [],
          },
        },
      },
    ]
    const { container } = render(<ChooseProfession />, withResultsMock)

    await wait()

    expect(container).toMatchSnapshot()
  })

  it('renders loading message', async () => {
    const { getByPlaceholderText, getByText } = render(
      <ChooseProfession />,
      withResultsMock
    )

    fireEvent.change(getByPlaceholderText('Yrkesroll eller yrkesomr책de'), {
      target: { value: 'Systemutvecklare' },
    })

    expect(getByText(/Loading.../i)).toBeInTheDocument()
  })

  it('renders error message', async () => {
    const withErrorMock = [
      {
        request: {
          query: GET_ONTOLOGY_CONCEPTS,
          variables: {
            filter: 'Systemutvecklare',
            type: OntologyType.Occupation,
          },
        },
        result: {
          error: new Error('Error while trying to get concepts'),
        },
      },
    ]

    const { getByPlaceholderText, getByText } = render(
      <ChooseProfession />,
      withErrorMock
    )

    fireEvent.change(getByPlaceholderText('Yrkesroll eller yrkesomr책de'), {
      target: { value: 'Systemutvecklare' },
    })

    await wait()

    expect(getByText(/Some error.../i)).toBeInTheDocument()
  })

  it('should render with taxonomy query result', async () => {
    const { getByPlaceholderText, getByText } = render(
      <ChooseProfession />,
      withResultsMock
    )

    fireEvent.change(getByPlaceholderText('Yrkesroll eller yrkesomr책de'), {
      target: { value: 'Systemutvecklare' },
    })

    await wait()

    expect(getByText('Systemutvecklare')).toBeInTheDocument()
  })

  it('renders next button', async () => {
    const { getByText } = render(<ChooseProfession />)

    expect(getByText('N채sta')).toBeInTheDocument()
  })
})
