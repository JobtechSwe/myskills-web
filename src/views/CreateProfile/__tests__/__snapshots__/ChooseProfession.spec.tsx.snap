// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views/ChooseProfession renders empty result 1`] = `
<div>
  <div
    class="css-owzlzk"
  >
    <h3
      class="css-1h019s7"
      color="black"
      font-family="default"
      font-size="small"
    >
      YRKE
    </h3>
    <h1
      class="css-tkwwg6"
      color="black"
      font-family="default"
      font-size="small,medium,large"
    >
      Vad vill du jobba med?
    </h1>
    <input
      class="css-1k0e6iw"
      name="search"
      placeholder="Yrkesroll eller yrkesområde"
    />
    <div>
      Dina valda kompetenser:
    </div>
    <a
      href="/skapa-cv/kompetenser"
    >
      <button
        class="css-r327x7"
        font-family="default"
        font-size="small,medium"
      >
        Nästa
      </button>
    </a>
  </div>
</div>
`;
